src/github.com/matrix-org/dendrite/mediaapi/writers/download.go

```diff
@@ -202,13 +198,317 @@ func (r *downloadRequest) respondFromLocalFile(w http.ResponseWriter, absBasePat
+			return nil, &resErr
+		}
+
+		if mediaMetadata != nil {
+			// If we have a record, we can respond from the local file
+			return mediaMetadata, nil
+		}
+
+		// Note: if the result was 200, we shouldn't get here
+		switch activeRemoteRequestResult.Result {
+		case 404:
+			return nil, &util.JSONResponse{
+				Code: 404,
+				JSON: jsonerror.NotFound("File not found."),
+			}
+		case 500:
+			r.Logger.Error("Other goroutine failed to fetch the remote file.")
+			resErr := jsonerror.InternalServerError()
+			return nil, &resErr
+		default:
+			r.Logger.Error("Other goroutine failed to fetch the remote file.")
+			return nil, &util.JSONResponse{
+				Code: activeRemoteRequestResult.Result,
+				JSON: jsonerror.Unknown("Failed to fetch file from remote server."),
+			}
+		}
+	}
+
+	// No active remote request so create one
+	activeRemoteRequests.MXCToResult[mxcURL] = &types.RemoteRequestResult{
+		Cond: &sync.Cond{L: activeRemoteRequests},
+	}
+	return nil, nil
+}
+
+// getRemoteFile fetches the file from the remote server and stores its metadata in the database
+// Only the owner of the activeRemoteRequestResult for this origin and media ID should call this function.
+func (r *downloadRequest) getRemoteFile(absBasePath types.Path, maxFileSizeBytes types.FileSizeBytes, db *storage.Database, activeRemoteRequests *types.ActiveRemoteRequests) *util.JSONResponse {
+	// Wake up other goroutines after this function returns.
+	isError := true
+	var result int
+	defer func() {
+		if isError {
+			// If an error happens, the lock MUST NOT have been taken, isError MUST be true and so the lock is taken here.
+			activeRemoteRequests.Lock()
+		}
+		defer activeRemoteRequests.Unlock()
+		mxcURL := "mxc://" + string(r.MediaMetadata.Origin) + "/" + string(r.MediaMetadata.MediaID)
+		if activeRemoteRequestResult, ok := activeRemoteRequests.MXCToResult[mxcURL]; ok {
+			r.Logger.Info("Signalling other goroutines waiting for this goroutine to fetch the file.")
+			if result == 0 {
+				r.Logger.Error("Invalid result, treating as InternalServerError")
+				result = 500
+			}
+			activeRemoteRequestResult.Result = result
+			activeRemoteRequestResult.Cond.Broadcast()
+		}
+		delete(activeRemoteRequests.MXCToResult, mxcURL)
+	}()
+
+	finalPath, duplicate, resErr := r.fetchRemoteFile(absBasePath, maxFileSizeBytes)
+	if resErr != nil {
+		result = resErr.Code
+		return resErr
+	}
+
+	// NOTE: Writing the metadata to the media repository database and removing the mxcURL from activeRemoteRequests needs to be atomic.
+	// If it were not atomic, a new request for the same file could come in in routine A and check the database before the INSERT.
+	// Routine B which was fetching could then have its INSERT complete and remove the mxcURL from the activeRemoteRequests.
+	// If routine A then checked the activeRemoteRequests it would think it needed to fetch the file when it's already in the database.
+	// The locking below mitigates this situation.
```

Comment:
Hmm. Probably better that we discuss this in #dendrite-dev:matrix.org as it's a bit too complicated to discuss back and forth here I think.